src/workerd/api/http.c++

```diff
@@ -2048,6 +2049,36 @@ jsg::Promise<QueueResponse> Fetcher::queue(
   });
 }
 
+jsg::Promise<kj::Maybe<jsg::V8Ref<v8::Value>>> Fetcher::getRpcAndCallRemote(
+    jsg::Lock& js, kj::String method, jsg::Optional<kj::Array<jsg::Value>> args) {
+  auto& ioContext = IoContext::current();
+  // TODO(now): Just a reminder for the future, rename this operation name.
+  auto worker = getClient(ioContext, nullptr, "rpcTest"_kjc);
+  auto event = kj::refcounted<api::JsRpcServiceCustomEventImpl>(9);
+  auto eventRef = kj::addRef(*event);
+
+  return ioContext.awaitIo(js, worker->customEvent(kj::mv(eventRef)).attach(kj::mv(worker)),
+      [event=kj::mv(event), &ioContext, method=kj::mv(method), args=kj::mv(args)]
+      (jsg::Lock& js, WorkerInterface::CustomEvent::Result result) mutable {
+    // We read the rpc capability off the event to get around customEvent's return type limitation.
```

Comment:
Hmm, it seems problematic that we're doing all this after `customEvent()` returns. Generally, the convention is that methods of WorkerInterface return promises, and the promise resolves when the event is done. Various parts of the stack may assume that the client is no longer connected to the server at this point.

Take HTTP `request()`, for example. This returns `Promise<void>`. That promise does not resolve until the HTTP request and response are _done_ -- the response has been fully written. In the case of a WebSocket, it doesn't resolve until the connection has closed in both directions.

I think we need to keep that convention here or we risk creating a lot of UAFs.

Here's what I think you should do instead: When `JsRpcServiceCustomEventImpl` is constructed, it immediately creates the `JsRpcTarget::Client` from a `kj::Promise<JsRpcTarget::Client>`, which is itself the promise end of a promise-fulfiller pair. Once the real capability is known, the fulfiller is fulfilled with it, connecting the client to the server.

This is also advantageous in that it means the client JavaScript can immediately begin expressing calls on this capability without having to await anything, and without dropping and re-taking the isolate lock.

Reply:
Gave this a shot, might just be a capnp misunderstanding but there may also be meaningful differences here?

`getHttpService` doesn't use `customEvent` -- on the (rpc) server-side we're just creating the http service, setting it on the response, and then using that immediately (due to pipelining I suppose?).

In this case, upon receiving the `getJsRpcTarget` event on the server side, we have to dispatch a `customEvent()`, whose `CustomEventImpl::run()` will fulfill the capability promise (which we want to set on the response). My understanding of capnp is the method running on the server-side has to resolve before the rpc response is sent over, is that right? But in that case, wouldn't we have to wait for `customEvent()` to complete on the server side before we can send the response (including the capability) to the client?

Briefly: we're creating a CustomEventImpl on the client-side and another on the server-side, the client-side capability is fulfilled in `sendRpc()`, but the `GetJsRpcTargetRequest()` resolves when the server-side customEvent() completes. On the server's end, we're calling `setServer()` on the Context for the response, but this only gets sent to the client after `customEvent()` finishes, right? Doesn't this present a problem?

Reply:
Again, the trick here is that you don't wait for the `JsRpcTarget::Client` to be returned. You create it immediately, based on a promise-fulfiller, and then on the receiving end you fulfill that with the final object. Calls can then start passing through without waiting for the top-level call to return.

The comparison to `getHttpService` is somewhat misleading here, because `getHttpService` actually doesn't initiate an event -- it's calling `request()` on the returned service that actually delivers an event.

That said, I suppose you could also argue that `getJsRpcTarget` should actually behave like `getHttpService`, i.e., it doesn't actually start an event. The event starts when an actual RPC call is made. But that might lead to other problems when we get to v2... at that point, the original call returning might not be the end of the RPC session, as any returned capabilities hold it open until they are dropped.

Reply:
Something like this https://github.com/cloudflare/workerd/pull/1028/commits/d3a8a3c13b4ca4a6ff4e06559da8d9a0cec76d71? 

> Calls can then start passing through without waiting for the top-level call to return.

I'm still kind of concerned that the "real capability" resolves after `customEvent` does, if I `co_await` [this line](https://github.com/cloudflare/workerd/pull/1028/commits/d3a8a3c13b4ca4a6ff4e06559da8d9a0cec76d71#diff-5bad3bda3228f7d05119093c652cfbd721a115b9e6f9de49e93db9d47ea45dbfR21) (specifically the `callPromise`) rather than adding it as a waitUntil, then `JsRpcTargetImpl::call()` seems to never run.

Reply:
> Something like this https://github.com/cloudflare/workerd/commit/d3a8a3c13b4ca4a6ff4e06559da8d9a0cec76d71?

Yep. I left some comments.

> I'm still kind of concerned that the "real capability" resolves after customEvent does, if I co_await [this line](https://github.com/cloudflare/workerd/pull/1028/commits/d3a8a3c13b4ca4a6ff4e06559da8d9a0cec76d71#diff-5bad3bda3228f7d05119093c652cfbd721a115b9e6f9de49e93db9d47ea45dbfR21) (specifically the callPromise) rather than adding it as a waitUntil, then JsRpcTargetImpl::call() seems to never run.

Unfortunately your link there doesn't seem to work. But I think I know what you mean. To solve this you need to use `context.setPipeline()` on the server side of the RPC, so that the client can start sending calls to the returned object before the call actually returns. (This would go in the internal PR, I will comment there too.)